synapse/lib/httpapi.py

```diff
@@ -1240,3 +1246,188 @@ async def get(self):
 
         resp = await self.cell.getCoreInfoV2()
         return self.sendRestRetn(resp)
+
+class ExtApiHandler(StormHandler):
+    '''
+    /api/ext/.*
+    '''
+
+    storm_prefix = 'init { $request = $lib.cortex.httpapi.response($_http_request_info) }'
+
+    # Disables the etag header from being computed and set. It is too much magic for
+    # a user defined API to utilize.
+    def compute_etag(self):
+        return None
+
+    def set_default_headers(self):
+        self.clear_header('Server')
+
+    async def get(self, path):
+        return await self._runHttpExt('get', path)
+
+    async def head(self, path):
+        return await self._runHttpExt('head', path)
+
+    async def post(self, path):
+        return await self._runHttpExt('post', path)
+
+    async def put(self, path):
+        return await self._runHttpExt('put', path)
+
+    async def delete(self, path):
+        return await self._runHttpExt('delete', path)
+
+    async def patch(self, path):
+        return await self._runHttpExt('patch', path)
+
+    async def options(self, path):
+        return await self._runHttpExt('options', path)
+
+    async def _runHttpExt(self, meth, path):
+        core = self.getCore()
+        adef, args = await core.getHttpExtApiByPath(path)
+        if adef is None:
+            self.set_status(404)
+            self.sendRestErr('NoSuchPath', f'No Extended HTTP API endpoint matches {path}')
+            return await self.finish()
+
+        requester = ''
+        iden = adef.get("iden")
+        useriden = adef.get('owner')
+
+        if adef.get('authenticated'):
+
+            requester = await self.useriden()
+
+            if requester is None:
+                await self.reqAuthUser()
+                return
+
+            for pdef in adef.get('perms'):
+                if not await self.allowed(pdef.get('perm'), default=pdef.get('default')):
+                    return
+
+            if adef.get('runas') == 'user':
+                useriden = requester
+
+        storm = adef['methods'].get(meth)
+        if storm is None:
+            self.set_status(500)
```

Comment:
Should this be a 405 with an `Allow: ...` list of supported methods for this endpoint? https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405

Reply:
That's a good call and quick to change.